Implement alarm shelving

dpino opened the pull request:
This feature was pending to be implemented.

Depends on https://github.com/Igalia/snabb/pull/1022

dpino committed:
Implement alarm shelving

wingo commented on src/lib/yang/alarms.lua:
```diff
@@ -102,6 +117,8 @@ function build_summary (alarms)
       ret[severity] = entry
    end
+   local shelved_alarms = table_size(state.shelved_alarms.shelved_alarms)
+   if shelved_alarms > 0 then ret['shelved_alarms'] = shelved_alarms end
```
Better to implement a `table_is_empty(t)` predicate that can be O(1):
```lua
function table_is_empty(t) for k,v in pairs(t) do return false end return true end
```

wingo commented on src/lib/yang/alarms.lua:
```diff
@@ -387,6 +424,11 @@ function set_operator_state (key, args)
+   if args.state == 'shelved' then
+      shelve_alarm(key, alarm)
+   elseif args.state == 'un-shelved' then
+      unshelve_alarm(key, alarm)
```
Is there a third case here?  If not, we should assert that state is either shelved or unshelved, or otherwise use a shelved boolean property.

dpino replied:
Besides 'shelved' and 'un-shelved' other operator states are 'none', 'ack' and 'closed'.

wingo reviewed (approved):
lgtm, does it work via the command line?

dpino committed:
Add failsafe parameter to common.parse_args

In the case a command that relies on the common.parse_args function
needs to do additional parsing, the failsafe mode prevents the
common.parse_args function from failing if there are leftover arguments
after parsing

dpino committed:
Refactor argument parsing

dpino committed:
Fix error message

dpino committed:
Update README

dpino committed:
Implement table_is_empty

dpino commented:
Test:

```bash
$ sudo ./snabb lwaftr run --name lwaftr --conf lwaftr.conf --on-a-stick 82:00.0
ARP: Resolving '10.0.1.2'
ARP: Resolving '10.0.1.2'
```

The lwaftr instance cannot resolve a next-hop IP address. Query alarms:

```
$ sudo ./snabb alarms get-state lwaftr /
  alarm-list {                              
    alarm {                                 
      alarm-type-id arp-resolution;         
      alarm-type-qualifier '';              
      resource 14472;                       
      alarm-text "Make sure you can resolve external-interface.next-hop.ip address manually.  If it cannot be resolved, consider setting the MAC address of the next-hop directly.  To do it so, set external-interface.next-hop.mac to the value of the MAC address.";
      is-cleared false;                     
      last-changed 2018-04-20T09:48:57Z;    
      operator-state-change {               
        time 2018-04-20T09:49:37Z;          
        operator admin;                     
        state ack;                          
        text '';                            
      }                                     
      perceived-severity critical;          
      status-change {                       
        time 2018-04-20T09:48:57Z;          
        alarm-text "Make sure you can resolve external-interface.next-hop.ip address manually.  If it cannot be resolved, consider setting the MAC address of the next-hop directly.  To do it so, set external-interface.next-hop.mac to the value of the MAC address.";
        perceived-severity critical;        
      }                                     
      time-created 2018-04-20T09:48:57Z;    
    }                                       
    last-changed 2018-04-20T09:48:57Z;      
    number-of-alarms 1;                     
  }
...                                         
```

Change operator-state of resource 14472 to ack:

```bash
$ sudo ./snabb alarms set-operator-state --schema snabb-softwire-v2 lwaftr 14472 arp-resolution ack
true
```

dpino commented:
I pushed more commits into this already reviewed PR. I refactored argument parsing for alarms subcommands. They rely in the config/common.lua tool for argument parsing (for reading values of flags such as `schema`, `revision`, etc. But since they can take additional arguments than the standard config commands (`get`, `set`, etc), it's necessary to do further parsing. For this reason, I added a `failsafe` flag to common, so if there are remaining arguments the parsing function doesn't fail.

wingo commented on src/program/config/common.lua:
```diff
@@ -24,6 +24,7 @@ local parse_command_line_opts = {
    require_schema = { default=false },
    is_config = { default=true },
    usage = { default=show_usage },
+   failsafe = { default=false },
```
Can we rename to "allow_extra_args" ?

wingo reviewed (approved):
lgtm with one nit

dpino committed:
Rename 'failsafe' to 'allow_extra_args'
